components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java

```diff
@@ -105,28 +105,30 @@ public static ApplicationManagementServiceImpl getInstance() {
     }
 
     @Override
-    public int createApplication(ServiceProvider serviceProvider, String tenantDomain, String userName)
+    public void createApplication(ServiceProvider serviceProvider, String tenantDomain, String userName)
             throws IdentityApplicationManagementException {
-        try {
 
-            startTenantFlow(tenantDomain, userName);
+        startTenantFlow(tenantDomain, userName);
 
-            // invoking the listeners
-            List<ApplicationMgtListener> listeners = ApplicationMgtListenerServiceComponent.getListners();
+        // invoking the listeners
+        List<ApplicationMgtListener> listeners = ApplicationMgtListenerServiceComponent.getListners();
 
-            for (ApplicationMgtListener listener : listeners) {
-                listener.createApplication(serviceProvider);
+        for (ApplicationMgtListener listener : listeners) {
+            if (!listener.doPreCreateApplication(serviceProvider)) {
```

Comment:
We may need to change the API for the listener to take tenant domain and username (as arguments) similar to the createApplication method in ApplicationManagementServiceImpl. Ideally the listener interfaces should be same as the service interface. Currently the listener doesn't get the tenant domain and username. It is however set in CarbonContext because we are starting the tenant flow. However when it comes to the post listeners further down you can see we have already ended the tenant flow, and now there is no way to find the tenant domain or username. So basically the API for the listener is wrong, and we need to fix it.